Insert transactions endpoint (Transaction monitoring)

MatanYadaev opened the pull request:
## **Type**
enhancement, bug_fix


___

## **Description**
- Simplified the generation of fake alert definitions and transactions, including support for business and end user as counterparties.
- Enhanced end user and business generation scripts with better handling of optional parameters and correlation IDs.
- Fixed a typo in `BusinessCreateDto`.
- Implemented bulk transaction creation endpoint and updated transaction creation logic.
- Refactored transaction mapping and repository logic to align with schema changes.
- Added integration tests for transaction creation endpoints.
- Updated Prisma schema and added a database migration to adjust unique constraints and remove unused enum.


___



## **Changes walkthrough**
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement
</strong></td><td><details><summary>8 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>generate-alerts.ts</strong><dd><code>Simplify Fake Alert Definitions Generation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

services/workflows-service/scripts/alerts/generate-alerts.ts
<li>Simplified the structure for generating fake alert definitions.<br> <li> Changed the <code>ids</code> parameter type to <code>string[]</code> for simplicity.<br> <li> Updated the fake alert generation logic to use simplified <code>ids</code>.<br>


</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2171/files#diff-65f82c5656fd8455fd436980f8c7d07ace7862374973f5493734b25cb69cce9f">+5/-31</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>generate-transactions.ts</strong><dd><code>Support Business and End User as Transaction Counterparties</code></dd></summary>
<hr>

services/workflows-service/scripts/alerts/generate-transactions.ts
<li>Added support for generating transactions with either business or end <br>user as counterparties.<br> <li> Removed the <code>CounterpartyType</code> usage following its removal from the <br>schema.<br> <li> Updated transaction generation to support new counterparty logic.<br>


</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2171/files#diff-33ff9e71663b769f6c500929e5b94fee94e61a1df7f4258daab6d2d8562203c0">+20/-23</a>&nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>generate-end-user.ts</strong><dd><code>Enhance End User and Business Generation Scripts</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

services/workflows-service/scripts/generate-end-user.ts
<li>Made <code>id</code> parameter optional in <code>generateEndUser</code> and <code>generateBusiness</code> <br>functions.<br> <li> Added <code>correlationId</code> handling in business and end user generation.<br> <li> Updated workflow parameter handling in <code>generateEndUser</code> to be optional.<br> <br>


</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2171/files#diff-a97db163336ad2d66d487b0efaa340cdc1b3acbd0c3f63af1234d6c47d2494f8">+30/-22</a>&nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>seed.ts</strong><dd><code>Update Seeding Script with End User Transactions</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

services/workflows-service/scripts/seed.ts
<li>Added creation of end users and associated transactions in the seeding <br>script.<br> <li> Updated transaction generation to include end users as counterparties.<br> <br>


</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2171/files#diff-f1d8de203cd513c3d62628f31af33153ed7a484b71329717c99bd0070cbbd7a3">+29/-3</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>transaction.controller.external.ts</strong><dd><code>Implement Bulk Transaction Creation Endpoint</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

services/workflows-service/src/transaction/transaction.controller.external.ts
<li>Implemented bulk transaction creation endpoint.<br> <li> Updated transaction creation to use the new bulk creation logic.<br>


</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2171/files#diff-7f26bfa0ccf29a96c0ffd9edae350cfd67a33bd51637e539f06143ed159e9cb8">+54/-24</a>&nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>transaction.mapper.ts</strong><dd><code>Refactor Transaction Mapping Logic</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

services/workflows-service/src/transaction/transaction.mapper.ts
<li>Refactored transaction mapping to support new counterparty logic.<br> <li> Removed unused mapping logic following schema changes.<br>


</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2171/files#diff-1060853f6596245e0ed9782bc95f9e7093d20d3d151e83cfe2056bab3317161f">+100/-105</a></td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>transaction.repository.ts</strong><dd><code>Update Transaction Repository for Schema Changes</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

services/workflows-service/src/transaction/transaction.repository.ts
<li>Removed deprecated fields and validation logic following schema <br>changes.<br> <li> Updated transaction creation logic to align with new schema.<br>


</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2171/files#diff-f3f6617a26249c990ea08ba2d5d922231dd4c586134c063c0f73151c1d29aef8">+2/-34</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>transaction.service.ts</strong><dd><code>Implement Bulk Transaction Creation Logic</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

services/workflows-service/src/transaction/transaction.service.ts
<li>Implemented logic for bulk transaction creation.<br> <li> Removed old batch creation logic.<br>


</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2171/files#diff-e3a24530c5cfef7de924d0eed08f4ab4a9864f3a7140ec551e8dbb885452f582">+25/-74</a>&nbsp; </td>
</tr>                    
</table></details></td></tr><tr><td><strong>Bug_fix
</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>business-create.ts</strong><dd><code>Fix Typo in BusinessCreateDto</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

services/workflows-service/src/business/dtos/business-create.ts
- Fixed typo in `correlationId` property of `BusinessCreateDto`.



</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2171/files#diff-d771e706bc2777885a447f3a8bae67f0604ca9185c237d3b7be061ce0905faa9">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    
</table></details></td></tr><tr><td><strong>Tests
</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>transaction.controller.external.intg.test.ts</strong><dd><code>Add Transaction Creation Integration Tests</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

services/workflows-service/src/transaction/transaction.controller.external.intg.test.ts
<li>Added integration tests for transaction creation endpoints, including <br>bulk creation.<br>


</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2171/files#diff-133ea78a07500a44ba8129bfac104ac74b7dbed45e6d23e2e28403d344995a37">+535/-0</a>&nbsp; </td>
</tr>                    
</table></details></td></tr><tr><td><strong>Configuration changes
</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>migration.sql</strong><dd><code>Database Migration for Schema Adjustments</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

services/workflows-service/prisma/migrations/20240305125152_fix_correlation_unique_and_remove_tm_denormalization/migration.sql
<li>Removed unique constraints on <code>transactionCorrelationId</code> in <br><code>TransactionRecord</code>.<br> <li> Added unique indexes on <code>projectId</code> and <code>correlationId</code> in <code>Business</code> and <br><code>EndUser</code>.<br> <li> Removed <code>CounterpartyType</code> enum and related column.<br>


</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2171/files#diff-bdc5871dec81738542a52789b36dfb77d57f890e794bbd95b6ac3a43b78795bf">+17/-0</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>schema.prisma</strong><dd><code>Prisma Schema Adjustments for Counterparty Handling</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

services/workflows-service/prisma/schema.prisma
<li>Removed unique constraint on <code>correlationId</code> for <code>EndUser</code> and <code>Business</code>.<br> <li> Added composite unique indexes for <code>correlationId</code> and <code>projectId</code> in <br><code>EndUser</code> and <code>Business</code>.<br> <li> Removed <code>CounterpartyType</code> enum and related fields.<br>


</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2171/files#diff-79f9532cd1d8e6c67881328fa550f851c8429f183de839b32575e959cdabef26">+10/-24</a>&nbsp; </td>
</tr>                    
</table></details></td></tr></tr></tbody></table>

___

> ✨ **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

MatanYadaev committed:
chore(.gitmodules): update ignore value for workflows-service data-migrations submodule to 'all' for consistency
chore(workflows-service): update data-migrations submodule to latest commit e2be2ce2e82641da25de172577c6835678dd2182

MatanYadaev committed:
feat(alerts): add support for generating fake alert definitions with configurable IDs
feat(alerts): add support for generating fake alert definitions with configurable severity and IDs
feat(alerts): add support for generating fake alert definitions with configurable execution details
feat(alerts): add support for generating fake transactions with configurable business and end user IDs
fix(schema.prisma): remove unique constraint on correlationId in EndUser and Business models
fix(schema.prisma): remove unique constraint on transactionCorrelationId in TransactionRecord model
fix(schema.prisma): remove unnecessary CounterpartyType enum
fix(migration.sql): remove unique constraint on correlationId in TransactionRecord model
fix(migration.sql): add unique index on projectId and correlationId in Business model
fix(migration.sql): add unique index on projectId and correlationId in EndUser model
fix(migration.sql): remove unnecessary Counterparty type and column in Counterparty model

feat(end-user): add support for creating end users with associated business

- Added `correlationId` field to the `generateBusiness` function in the `generate-end-user.ts` script to generate a correlation ID for each business.
- Modified the `generateBusiness` function in the `generate-end-user.ts` script to include the `correlationId` parameter in the function signature and assign it to the `correlationId` field of the generated business.
- Modified the `generateEndUser` function in the `generate-end-user.ts` script to include the `correlationId` parameter in the function signature and assign it to the `correlationId` field of the generated end user.
- Modified the `generateEndUser` function in the `generate-end-user.ts` script to make the `workflow` parameter optional.
- Modified the `generateEndUser` function in the `generate-end-user.ts` script to check if `workflowDefinitionId` is present before creating `workflowRuntimeData`.
- Modified the `seed` function in the `seed.ts` script to create two additional end users with associated transactions.
- Created a new file `end-user-create-with-business.ts` in the `end-user/dtos` directory to define the `EndUserCreateWithBusinessDto` class, which extends the `EndUserCreateDto` class and includes the `companyName` field.
- Modified the `EndUserControllerExternal` class in the `end-user.controller.external.ts` file to import and use the `EndUserCreateWithBusinessDto` class for the `createWithBusiness` endpoint.
- Removed the `findByCorrelationIdUnscoped` method from the `EndUserRepository` class in the `end-user.repository.ts` file as it is no longer used.
- Modified the `EndUserService` class in the `end-user.service.ts` file to remove the `correlationId` field from the `endUser` parameter in the `createWithBusiness` method.

MatanYadaev committed:
feat(transaction.controller.external.intg.test.ts): add integration tests for TransactionControllerExternal

This commit adds integration tests for the TransactionControllerExternal in the transaction module. The tests cover the following scenarios:

1. Creating a transaction with a business originator and an end user beneficiary.
2. Creating a transaction and reusing an existing business as the originator and an existing end user as the beneficiary.
3. Returning a 400 status code when validation fails.

The tests use the supertest library to send HTTP requests to the server and assert the responses. The tests also use the faker-js library to generate random data for the transactions.

The tests are organized into the following sections:
1. Setup: This section sets up the necessary dependencies and initializes the Nest application.
2. POST /: This section tests the creation of a single transaction.
3. POST /bulk: This section tests the creation of multiple transactions.

The tests assert the status code and the response body of the HTTP requests. They also assert the data stored in the database after the transactions are created.

MatanYadaev committed:
fix(business-create.ts): correct spelling of 'correlationId' property name
fix(data-analytics.service.intg.test.ts): remove unused import and CounterpartyType usage
fix(end-user-create.ts): remove unused properties and add required decorator to 'correlationId' property
feat(bulk-transactions-created.dto.ts): create new file for BulkTransactionsCreatedDto class
fix(get-transactions.dto.ts): remove unused 'businessId' property
fix(transaction-create.dto.ts): change 'date' property type to string and remove unused properties
feat(transaction-created.dto.ts): create new file for TransactionCreatedDto class

fix(transaction.controller.external.ts): fix import statements to improve readability and maintainability
feat(transaction.controller.external.ts): add validation pipe to validate request body in create method
feat(transaction.controller.external.ts): add support for bulk transaction creation in createBulk method
feat(transaction.controller.external.ts): handle creation responses and return appropriate status codes and response bodies in create and createBulk methods
refactor(transaction.controller.external.ts): remove unnecessary include in getTransactions method

feat(transaction): refactor transaction.mapper.ts to improve code readability and maintainability
fix(transaction.repository): remove unused code and simplify create method

feat(transaction.service.ts): add support for bulk transaction creation

- Add `createBulk` method to `TransactionService` to handle bulk creation of transactions.
- Update `create` method to use `TransactionEntityMapper` to map payload to entity.
- Remove unused imports and variables.

fix(transaction.service.ts): fix typo in import statement

- Fix typo in import statement for `TransactionCreatedDto`.

refactor(transaction.service.ts): improve error handling in bulk transaction creation

- Update `createBulk` method to handle errors during transaction creation.
- Return an array of `TransactionCreatedDto` or an object with `error` and `correlationId` properties for each transaction.

refactor(transaction.service.ts): remove unused import and file

- Remove unused import for `TProjectId`.
- Remove unused file `create-transaction.validator.ts`.

MatanYadaev commented on .gitmodules:
```diff
@@ -1,4 +1,4 @@
 	path = services/workflows-service/prisma/data-migrations
 	url = git@github.com:ballerine-io/wf-data-migration.git
-	ignore = dirty
+	ignore = all
```
Boyscouting: This flag should make `data-migration` submodule out of our PRs

MatanYadaev commented on services/workflows-service/prisma/schema.prisma:
```diff
@@ -759,8 +746,7 @@ enum SanctionListMatchStatus {
 
-  correlationId String?
-  type          CounterpartyType
+  correlationId String? // This column is a denormalized value of the EndUser.correlationId or Business.correlationId
```
Kept this column as a denormalization of EndUser.correlationId or Business.correlationId. Not ideal, but it helps for now, and as `Counterparty` is a model that does not update, it has relatively low risk from being inconsistent.

MatanYadaev commented on services/workflows-service/scripts/seed.ts:
```diff
@@ -146,19 +146,45 @@ async function seed() {
 
+  const endUser1 = await client.endUser.create({
+    data: generateEndUser({
+      projectId: project1.id,
+    }),
+  });
+
+  const endUser2 = await client.endUser.create({
+    data: generateEndUser({
+      projectId: project1.id,
+    }),
+  });
+
   const ids1 = await generateTransactions(client, {
     projectId: project1.id,
-    businessId: business1.id,
+    business: business1,
+    endUser: endUser1,
   });
   const ids2 = await generateTransactions(client, {
     projectId: project1.id,
-    businessId: business2.id,
+    business: business2,
+    endUser: endUser1,
+  });
+  const ids3 = await generateTransactions(client, {
+    projectId: project1.id,
+    business: business2,
+    endUser: endUser1,
+  });
+  const ids4 = await generateTransactions(client, {
+    projectId: project1.id,
+    business: business2,
+    endUser: endUser2,
```
Added some variety to our transactions seed:
1. Transaction with both originator & beneficiary that can be either business or end user
2. Transaction with originator only
3. Transaction with beneficiary only

MatanYadaev commented on services/workflows-service/src/end-user/end-user.repository.ts:
```diff
@@ -59,16 +59,6 @@ export class EndUserRepository {
     });
   }
 
-  async findByCorrelationIdUnscoped<T extends Omit<Prisma.EndUserFindUniqueArgs, 'where'>>(
```
Unused method

MatanYadaev commented on services/workflows-service/src/transaction/transaction.mapper.ts:
```diff
@@ -59,110 +45,119 @@ export class TransactionEntityMapper {
+                        },
+                      },
+                      create: {
+                        project: { connect: { id: projectId } },
+                        correlationId: counterpartyInfo.correlationId,
+                        companyName: counterpartyInfo.businessData.companyName,
+                        registrationNumber: counterpartyInfo.businessData.registrationNumber,
+                        mccCode: counterpartyInfo.businessData.mccCode,
+                        businessType: counterpartyInfo.businessData.businessType,
+                      },
+                    },
+                  }
+                : {},
+              endUser: counterpartyInfo.endUserData
+                ? {
+                    connectOrCreate: {
+                      where: {
+                        projectId_correlationId: {
+                          correlationId: counterpartyInfo.correlationId,
+                          projectId: projectId,
+                        },
+                      },
+                      create: {
+                        project: { connect: { id: projectId } },
+                        correlationId: counterpartyInfo.correlationId,
+                        firstName: counterpartyInfo.endUserData.firstName,
+                        lastName: counterpartyInfo.endUserData.lastName,
+                        email: counterpartyInfo.endUserData.email,
+                        isContactPerson: counterpartyInfo.endUserData.isContactPerson,
+                        phone: counterpartyInfo.endUserData.phone,
+                        dateOfBirth: counterpartyInfo.endUserData.dateOfBirth,
+                        avatarUrl: counterpartyInfo.endUserData.avatarUrl,
+                      },
+                    },
+                  }
+                : {},
+            },
+          },
+        }
+      : undefined;
```
Here we're upserting the counterparty and its entity. A bit long, but quite simple and covered with tests.

MatanYadaev committed:
feat(alerts): generate transactions for a project

- Generate transactions for a project using the `generateTransactions` function.
- Create counterparties and collect their IDs.
- For each counterparty, create a counterparty record with a correlation ID and connect it to the project.
- For each counterparty, create a business or end user record with a correlation ID and connect it to the counterparty.
- Update the number of iterations for creating counterparty records from 200 to 100.
- Update the number of iterations for generating transactions from 4 to 2.
- Update the `generateFakeAlertDefinitions` function to use the counterparty originator IDs from the generated transactions.

feat(seed): remove unnecessary business and end user creation

- Remove the creation of unnecessary business and end user records in the `seed` function.

MatanYadaev committed:
fix(transactions): remove unused businessId field from fetchers and useTransactionsQuery

The `businessId` field was removed from the `fetchers.ts` file in the `TransactionsListSchema` and `fetchTransactions` functions. This field was not being used and was causing unnecessary complexity.

In addition, the `businessId` field was also removed from the `useTransactionsQuery` function in the `useTransactionsQuery.tsx` file. This field was not being used and was unnecessary for the functionality of the query.

These changes were made to simplify the code and remove unused variables.

liorzam commented on services/workflows-service/src/transaction/transaction.mapper.ts:
```diff
@@ -59,110 +45,119 @@ export class TransactionEntityMapper {
+    | Prisma.TransactionRecordCreateInput['counterpartyOriginator']
+    | Prisma.TransactionRecordCreateInput['counterpartyBeneficiary'] {
+    if (counterpartyInfo?.businessData && counterpartyInfo?.endUserData) {
+      throw new Error('Counterparty must have either business or end user data, not both.');
```
its should throw 400 validation error, isnt?

MatanYadaev replied:
Correct, I'll fix it

liorzam commented on services/workflows-service/src/transaction/transaction.controller.external.ts:
```diff
@@ -24,38 +35,62 @@ export class TransactionControllerExternal {
-    if (batchCreateResponse.overallStatus === 'partial') {
-      response.status(207);
-    }
+    const creationResponses = await this.service.createBulk({
+      transactionsPayload: body,
+      projectId: currentProjectId,
+    });
```
we already agreed on bulk response,  which indicates for each row if had an issue,

error response for bulk for example: `Alerts - Bulk assign - wrong alert id` API: `/api/v1/external/alerts/decision` 
```
 
{
    "overallStatus": "failed",
    "response": [
        {
            "alertId": "clshptjjr03pipgdoq7108",
            "status": "failed",
            "error": {
                "message": "Alert not found or not updated."
            }
        }
    ]
}
```

liorzam replied:
Also success should be look like :
```
{
    "overallStatus": "success",
    "response": [
        {
            "alertId": "clshptjjr03pipgdoq7108hip",
            "status": "success"
        }
    ]
}
```

liorzam commented on services/workflows-service/src/transaction/transaction.service.ts:
```diff
@@ -14,86 +14,37 @@ export class TransactionService {
-          txid: res.id,
-          txCorrelationId: transaction.correlationId,
-          status: 'success',
+        const transaction = await this.repository.create({
+          data: TransactionEntityMapper.toCreateData({
+            dto: transactionPayload,
+            projectId,
+          }),
         });
```
How do we prevent from duplication of transaction from being inserted?

MatanYadaev replied:
`correlationId` is unique, it will fail if you'll try inserting it twice.

liorzam commented on services/workflows-service/src/transaction/transaction.service.ts:
```diff
@@ -14,86 +14,37 @@ export class TransactionService {
+          }),
         });
-      } catch (error) {
-        overallStatus = 'partial';
```
this is exactly part of our bulk response you've just removed 
overallStatus: success|failed|partial

MatanYadaev replied:
Why is it important?
We return 201 if it was successful, and 209 for partial. I can return 500 if all of the all of the insertions were failed.

liorzam commented on services/workflows-service/src/transaction/dtos/transaction-create.dto.ts:
```diff
@@ -31,16 +30,14 @@ import { EndUserCreateDto } from '@/end-user/dtos/end-user-create';
+  businessData?: Omit<BusinessCreateDto, 'correlationId'>;
+
+  @IsOptional()
+  endUserData?: Omit<EndUserCreateDto, 'correlationId'>;
```
Both are optional?
Dont we want to declare that one of them have to present?

MatanYadaev replied:
As I understood, both can be optional, and I wrote a test for it. I'll verify it with Alon.

MatanYadaev committed:
fix(alert.controller.external.ts): fix import statement for express.Response
feat(alert.controller.external.ts): add support for returning 207 status code for partial success in assignAlertById and decision methods
fix(assign-alert.dto.ts): remove unused imports and update AlertUpdateResponse class to include errors property
fix(types.ts): remove PARTIAL status from BulkStatus enum and update TBulkAssignAlertsResponse type to remove overallStatus property

fix(sentry.interceptor.ts): refactor SentryInterceptor to use SentryService for capturing exceptions and http errors
feat(sentry.service.ts): create SentryService to handle capturing exceptions and http errors with additional data
fix(bulk-transactions-created.dto.ts): add optional error field to BulkTransactionsCreatedDto
fix(transaction-create.dto.ts): make originator and beneficiary fields required in TransactionCreateDto
fix(transaction-created.dto.ts): make id field optional in TransactionCreatedDto

fix(transaction.controller.external.intg.test.ts): remove unused import of TransactionRepository
feat(transaction.controller.external.intg.test.ts): refactor test cases to use constants for bulk status and improve readability
fix(transaction.controller.external.intg.test.ts): fix validation test case to use valid transaction data for all transactions

fix(transaction.controller.external.ts): import missing BulkStatus and errors modules to resolve compilation errors
feat(transaction.controller.external.ts): add additional swagger annotations to createBulk endpoint for improved documentation
fix(transaction.mapper.ts): throw HttpException instead of Error when counterparty has both business and end user data
feat(transaction.module.ts): add SentryService to providers array for error tracking
feat(transaction.service.ts): add SentryService to constructor and capture exceptions for error tracking

MatanYadaev committed:
feat(alert.module.ts): add SentryModule to the imports to enable error tracking with Sentry
fix(assign-alert.dto.ts): add enum validation for the status property in the AlertUpdateResponse class to ensure it only accepts valid values from the BulkStatus enum
fix(swagger.ts): fix typo in the Sandbox Server URL in the addServer method
fix(bulk-transactions-created.dto.ts): add enum validation for the status property in the BulkTransactionsCreatedDto class to ensure it only accepts valid values from the BulkStatus enum
feat(transaction.controller.external.ts): add ApiBody decorator to the createBulk method to specify the request body type as an array of TransactionCreateDto

liorzam commented on services/workflows-service/src/transaction/transaction.controller.external.intg.test.ts:
```diff
@@ -28,9 +28,53 @@ import { TransactionModule } from '@/transaction/transaction.module';
 
-const getBaseTransactionData = (): TransactionCreateDto => {
+const getBusinessCounterpartyData = (business?: Business) => {
+  if (business) {
+    return {
+      correlationId: business.correlationId!,
+      businessData: {
+        companyName: business.companyName,
+        registrationNumber: business.registrationNumber!,
+        mccCode: business.mccCode!,
+        businessType: business.businessType!,
+      },
+    };
+  }
+
+  return {
+    correlationId: faker.datatype.uuid(),
+    businessData: {
+      companyName: faker.company.name(),
+      registrationNumber: faker.datatype.uuid(),
+      mccCode: faker.datatype.number({ min: 1000, max: 9999 }),
+      businessType: faker.lorem.word(),
+    },
+  };
```
could be only one return statement

MatanYadaev replied:
How will it make the code better?

liorzam commented on services/workflows-service/src/transaction/transaction.service.ts:
```diff
@@ -37,8 +40,16 @@ export class TransactionService {
+        if (error instanceof Prisma.PrismaClientKnownRequestError && error.code === 'P2002') {
+          errorToLog = new Error('Transaction already exists', { cause: error });
+        } else {
+          this.sentry.captureException(errorToLog);
```
it would be better just to capture the original error instead yours

MatanYadaev replied:
As long as I add the original error to the chain (`cause`), why does it matter?

liorzam commented on services/workflows-service/src/transaction/transaction.service.ts:
```diff
@@ -37,8 +40,16 @@ export class TransactionService {
       } catch (error) {
+        let errorToLog: Error = new Error('Unknown error', { cause: error });
+
+        if (error instanceof Prisma.PrismaClientKnownRequestError && error.code === 'P2002') {
```
you can use some prisma util functions like 'isFkConstraintError' which located at `services/workflows-service/src/prisma/prisma.util.ts`

MatanYadaev replied:
When will this occur?

liorzam commented on services/workflows-service/src/transaction/dtos/transaction-create.dto.ts:
```diff
@@ -33,10 +34,14 @@ export class CounterpartyInfo {
+  @ApiProperty({ type: OmitType(EndUserCreateDto, ['correlationId']) })
+  @ValidateIf(obj => !obj.businessData)
+  @ValidateNested()
   endUserData?: Omit<EndUserCreateDto, 'correlationId'>;
```
Can be implemented by ValidatorConstraintInterface which allow at least one field to be present
https://github.com/typestack/class-validator/issues/1581



```
// at-least-one.decorator.ts
import { ValidationArguments, ValidatorConstraint, ValidatorConstraintInterface } from 'class-validator';

@ValidatorConstraint({ name: 'atLeastOne', async: false })
export class AtLeastOneConstraint implements ValidatorConstraintInterface {
    validate(value: any, args: ValidationArguments) {
        const { object } = args;
        const correlatedFields = args.constraints[0];
        return correlatedFields.some(field => object[field] !== undefined);
    }

    defaultMessage(args: ValidationArguments) {
        return `At least one of the fields ${args.constraints[0].join(', ')} must be present`;
    }
}
```

liorzam replied:
```
import { AtLeastOneConstraint } from './at-least-one.decorator';

export class MyDTO {
    @IsString()
    propertyA?: string;

    @IsString()
    propertyB?: string;

    @Validate(AtLeastOneConstraint, ['propertyA', 'propertyB'])
    atLeastOnePresent: string;
}
```

liorzam commented on services/workflows-service/src/transaction/dtos/transaction-create.dto.ts:
```diff
@@ -33,10 +34,14 @@ export class CounterpartyInfo {
   @ApiProperty({ required: false }) @IsString() @IsOptional() bankCountry?: string;
 
-  @IsOptional()
+  @ApiProperty({ type: OmitType(BusinessCreateDto, ['correlationId']) })
+  @ValidateIf(obj => !obj.endUserData)
+  @ValidateNested()
   businessData?: Omit<BusinessCreateDto, 'correlationId'>;
 
-  @IsOptional()
+  @ApiProperty({ type: OmitType(EndUserCreateDto, ['correlationId']) })
+  @ValidateIf(obj => !obj.businessData)
+  @ValidateNested()
   endUserData?: Omit<EndUserCreateDto, 'correlationId'>;
 }
```
Can be implemented by ValidatorConstraintInterface which allow at least one field to be present
https://github.com/typestack/class-validator/issues/1581

```
import { ValidationArguments, ValidatorConstraint, ValidatorConstraintInterface } from 'class-validator';

@ValidatorConstraint({ name: 'atLeastOne', async: false })
export class AtLeastOneConstraint implements ValidatorConstraintInterface {
    validate(value: any, args: ValidationArguments) {
        const { object } = args;
        if (object.endUserData !== undefined) {
            object.atLeastOnePresent = object.endUserData;
        } else if (object.businessData !== undefined) {
            object.atLeastOnePresent = object.businessData;
        }
        return object.endUserData !== undefined || object.businessData !== undefined;
    }

    defaultMessage(args: ValidationArguments) {
        return 'At least one of endUserData or businessData must be present';
    }
}
```

MatanYadaev committed:
fix(transaction): throw HttpException when originator or beneficiary is missing in TransactionEntityMapper
fix(transaction): throw HttpException when counterparty has neither business nor end user data in TransactionEntityMapper
fix(transaction): throw HttpException when counterparty has both business and end user data in TransactionEntityMapper
fix(transaction.service): map transactionsPayload to mappedTransactions before creating transactions in bulk

MatanYadaev committed:
fix(HttpExceptions.filter.ts): import PRISMA_UNIQUE_CONSTRAINT_ERROR constant from prisma.util to handle unique key constraint violation error
fix(prisma.util.ts): add PRISMA_UNIQUE_CONSTRAINT_ERROR constant to represent the error code for unique key constraint violation
fix(transaction.service.ts): import PRISMA_UNIQUE_CONSTRAINT_ERROR constant from prisma.util to handle unique key constraint violation error

MatanYadaev committed:
refactor(transaction.controller.external.intg.test.ts): reformat code for better readability and maintainability
test(transaction.controller.external.intg.test.ts): add test case for when counterparty has both business and end user data, returns 400 status code and appropriate error message
